### PR TITLE
Better addon apis & devkits for addons

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: creating addons is too complicated

## What I did
With some new apis based on react hooks, this can be made much easier!

Our addons in the monorepo have grown to quite some complexity over time, because of backwards compatibility, feature creep of just because of "shit happens".

Because of this, it can be hard to point to one of our addons as an example on how to write addons.

This PR introduces "devkits", which are like tiny featureless addons that demonstrate a particular variant/type of addon.

We could move those devkits into their own repo as well, to make it easier for users to begin development of their own addon.

### CODE REVIEW

**Comment by Hypnosphi** [comment]
> based on react hooks

Does it mean dropping support for React <16.8?

**Comment by ndelangen** [comment]
These hooks are used solely in the manager @Hypnosphi

It does mean the manager's react version is locked to the high version indeed, but that version has no relationship with the version the user is using in the preview.

**Comment by UsulPro** [comment]
Hi Norbert! 
Hi guys! 

Nice to see steps towards simplifying addon creation. Hope it'll encourage the creation of a lot of new community addons

Funny, but two weeks ago I decided to continue my long term work (with the same name: `addon-devkit`). Basically it's the base which can be used for addon creation. It solves the problem of communication between manager and preview sides: instead of sending/listening for data, you just connect your addon/decorator to the addon store like you do with redux store.

Finally, here is v1.0.0
https://github.com/storybooks/addon-development-kit

I created it keeping in mind some of my addon ideas. But I believe it's pretty universal and could be helpful for many cases. So what do you think about it in the relation of the topic of this PR?

**Comment by Hypnosphi** [comment]
@ndelangen Ok, then we really need to raise the react version from 16.7 to 16.8 in `@storybook/api` and other manager packages. Otherwise they can break on user side because of npm deduping

**Comment by ndelangen** [comment]
> need to raise the react version from 16.7 to 16.8

I'll do so, thanks for the heads-up

Would you like to pair program some time on this "hooks api in the preview"?

**Comment by ndelangen** [comment]
@UsulPro let's discuss on discord? I love the project goals!

First I want to lay a good foundation of lower level APIs, then figure out good abstractions on top.

**Comment by Hypnosphi** [comment]
> Would you like to pair program some time on this "hooks api in the preview"?

Sure, but probably not these two weeks

File: lib/api/src/index.tsx [hunk 1]
```diff
@@ -268,5 +271,61 @@ export function useStorybookState(): State {
...
+  return api;
+}
 
 export { ManagerConsumer as Consumer, ManagerProvider as Provider };
+
+export interface EventMap {
+  [eventId: string]: Listener;
+}
+
+function itOrDefault<S>(fromStore: S, defaultState: S): S {
```
Comment (shilman) [inline lib/api/src/index.tsx]: `orDefault`?

File: lib/components/src/addon-panel/addon-panel.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  useEffect(() => {
+    // happens after return
+    ref.current = value;
+  }, [value]);
+
+  return ref.current;
+};
+
+const useUpdate = (update: boolean, value: any) => {
```
Comment (shilman) [inline lib/components/src/addon-panel/addon-panel.tsx]: Comment these? When/how are they supposed to be used?

**Review by shilman — APPROVED** [review]
Looks great. Left a bunch of comments inline, none of them particularly crucial other than the documentation issues, which I believe is what this is supposed to address.

Are you ok with delaying merge until after 5.1 ships?

File: lib/components/src/addon-panel/addon-panel.tsx (hunk 2 above)
Comment (ndelangen) [reply lib/components/src/addon-panel/addon-panel.tsx]: it's not exported, so it's just for use in this one place for now.

File: addons/design-assets/src/panel.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+  if (results.length && !results[selected]) {
+    setSelected(0);
+    return null;
+  }
+
+  return useMemo(() => {
+    const url = getUrl(results[selected]).replace('{id}', storyId);
+    return (
+      <Fragment>
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply addons/design-assets/src/panel.tsx]: I'll pass on using `<>`

File: dev-kits/addon-roundtrip/src/panel.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import React, { Fragment } from 'react';
+import { useAddonState, useChannel } from '@storybook/api';
+import { ActionBar } from '@storybook/components';
+import { ADDON_ID, EVENTS } from './constants';
+
+type Results = string[];
+
+interface ContentProps {
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply dev-kits/addon-roundtrip/src/panel.tsx]: yes, the types can be removed simple enough to reduce it into plain JS I think

File: dev-kits/addon-roundtrip/src/panel.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+));
+
+export const Panel = () => {
+  const [results, setState] = useAddonState<Results>(ADDON_ID, []);
+  const emit = useChannel({
+    [EVENTS.RESULT]: (newResults: Results) => setState(newResults),
+  });
+
+  return (
+    <Fragment>
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply dev-kits/addon-roundtrip/src/panel.tsx]: no thanks

File: dev-kits/addon-parameter/README.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# example addon: parameter
+
+This addon listens to changes in story parameters, and displays them
+
+This can be useful to display metadata on component/stories
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply dev-kits/addon-parameter/README.md]: I don't really want to link to addons that have the same concept but use totally different APIs, I fear that will just lead to confusion tbh.

**Pull request merged**